### PR TITLE
Support lightstep.BinaryCarrier

### PR DESCRIPTION
**Pull request by jmacd**

### CODE REVIEW

File: basictracer/propagation_ls.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		return err
+	}
+	if bcarrier == nil {
+		*scarrier = base64.StdEncoding.EncodeToString(data)
+	} else {
+		*bcarrier = make([]byte, base64.StdEncoding.EncodedLen(len(data)))
+		base64.StdEncoding.Encode(*bcarrier, data)
+	}
+	return nil
+}
```
Comment (tedsuo) [inline basictracer/propagation_ls.go]: Small readability/optimization improvement: please consider this version of `Inject` that has fewer conditionals and extra allocations (and something similar for `Extract`).
```go
{
	sc, ok := spanContext.(SpanContext)
	if !ok {
		return opentracing.ErrInvalidSpanContext
	}

	data, err := proto.Marshal(&lightstep.BinaryCarrier{
		BasicCtx: &lightstep.BasicTracerCarrier{
			TraceId:      sc.TraceID,
			SpanId:       sc.SpanID,
			Sampled:      true,
			BaggageItems: sc.Baggage,
		}
	})
	if err != nil {
		return err
	}

	switch carrier := opaqueCarrier.(type) {
	case *string:
		*carrier = base64.StdEncoding.EncodeToString(data)
	case *[]byte:
		*carrier = make([]byte, base64.StdEncoding.EncodedLen(len(data)))
		base64.StdEncoding.Encode(*carrier, data)
	default:
		return opentracing.ErrInvalidCarrier
	}

	return nil
}
```

File: basictracer/propagation_ls.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+type lightstepBinaryPropagator struct{}
+
+type lightstepBinaryCarrier struct{}
+
+var BinaryCarrier lightstepBinaryCarrier
+
+func (_ *lightstepBinaryPropagator) Inject(
```
Comment (jmacd) [inline basictracer/propagation_ls.go]: Done

File: basictracer/propagation_ls.go (hunk 1 above)
Comment (jmacd) [inline basictracer/propagation_ls.go]: Done

**Comment by jmacd** [comment]
I added a TODO about consolidating packages.

**Comment by RaduBerinde** [comment]
Hi @jmacd - quick question, what is the main motivation for the binary carrier? Is it expected to have less overhead than a `HTTPHeaders` carrier?  I'm trying to decide which one to use (I'm leaning towards `HTTPHeaders` so I can use the grpc-opentracing interceptors unmodified).

**Comment by jmacd** [comment]
Yes, the BinaryCarrier is partly about efficiency, but also provides a path
for forward/backward compatibility in case a Tracer implementation uses an
alternative representation for SpanContext.

I believe the grpc-opentracing plugin should be changed to recognize binary
carriers, actually, but that will take some carefully worded arguments and
may not happen soon. (It came up today in another context, so it's worth
thinking about.)

ᐧ

**Comment by RaduBerinde** [comment]
Thanks! Makes sense!

**Pull request merged**